File: Source/Fantasy.Jobs/Solar/SolarActionQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.ServiceModel;
using System.Threading;

namespace Fantasy.Jobs.Solar
{
    public class SolarActionQueue : AbstractService, ISolarActionQueue
    {

        private List<Action<ISolar>> _queue = new List<Action<ISolar>>();

        private Thread _retryThread;

        public override void InitializeService()
        {
            _retryThread = ThreadFactory.CreateThread(Refresh).WithStart();
            
            base.InitializeService();
        }


        private void Refresh()
        {
            while (true)
            {
                Thread.Sleep(15 * 1000);
                List<Action<ISolar>> list;
                lock (_queue)
                {
                    list = new List<Action<ISolar>>(_queue);
                }
                try
                {
                    using (ClientRef<ISolar> client = ClientFactory.Create<ISolar>())
                    {
                        foreach (Action<ISolar> action in list)
                        {
                            action(client.Client);
                            lock (_queue)
                            {
                                _queue.Remove(action);
                            }
                        }
                    }
                }
                catch
                {
                }

               
            }
        }

        public override void UninitializeService()
        {
            _retryThread.Abort();
            _retryThread.Join();
            base.UninitializeService();
        }


        public void Enqueue(Action<ISolar> action)
        {
            try
            {
                using (ClientRef<ISolar> client = ClientFactory.Create<ISolar>())
                {
                    action(client.Client);
                }
            }
            catch(Exception error)
            {
                if (WCFExceptionHandler.CanCatch(error))
                {
                    lock (_queue)
                    {
                        _queue.Add(action);
                    }
                }
                else
                {
                    throw;
                }
            }
        }
    }
}

File: branches/Stheno/Fantasy.Jobs/src/fantasy/jobs/solar/SolarActionQueue.java
package fantasy.jobs.solar;

import Fantasy.ServiceModel.*;

public class SolarActionQueue extends AbstractService implements ISolarActionQueue
{

	private java.util.ArrayList<Action<ISolar>> _queue = new java.util.ArrayList<Action<ISolar>>();

	private Thread _retryThread;

	@Override
	public void InitializeService()
	{
		_retryThread = ThreadFactory.CreateThread(Refresh).WithStart();

		super.InitializeService();
	}


	private void Refresh()
	{
		while (true)
		{
			Thread.sleep(15 * 1000);
			java.util.ArrayList<Action<ISolar>> list;
			synchronized (_queue)
			{
				list = new java.util.ArrayList<Action<ISolar>>(_queue);
			}
			try
			{
//C# TO JAVA CONVERTER NOTE: The following 'using' block is replaced by its Java equivalent:
//				using (ClientRef<ISolar> client = ClientFactory.Create<ISolar>())
				ClientRef<ISolar> client = ClientFactory.<ISolar>Create();
				try
				{
					for (Action<ISolar> action : list)
					{
						action(client.Client);
						synchronized (_queue)
						{
							_queue.remove(action);
						}
					}
				}
				finally
				{
				}
			}
			catch (java.lang.Exception e)
			{
			}


		}
	}

	@Override
	public void UninitializeService()
	{
		_retryThread.stop();
		_retryThread.join();
		super.UninitializeService();
	}


	public final void Enqueue(Action<ISolar> action)
	{
		try
		{
//C# TO JAVA CONVERTER NOTE: The following 'using' block is replaced by its Java equivalent:
//			using (ClientRef<ISolar> client = ClientFactory.Create<ISolar>())
			ClientRef<ISolar> client = ClientFactory.<ISolar>Create();
			try
			{
				action(client.Client);
			}
			finally
			{
			}
		}
		catch(RuntimeException error)
		{
			if (WCFExceptionHandler.CanCatch(error))
			{
				this.Site.<ILogger>GetService().SafeLogError("Solar", error, "WCF error");
				synchronized (_queue)
				{
					_queue.add(action);
				}
			}
			else
			{
				throw error;
			}
		}
	}
}